app/screens/manage_channel_members/manage_channel_members.tsx

```diff
@@ -0,0 +1,277 @@
+
+export default function ManageChannelMembers({
+    canManageAndRemoveMembers,
+    channelId,
+    componentId,
+    currentTeamId,
+    currentUserId,
+    teammateNameDisplay,
+    tutorialWatched,
+}: Props) {
+    const serverUrl = useServerUrl();
+    const theme = useTheme();
+    const {formatMessage} = useIntl();
+
+    const searchTimeoutId = useRef<NodeJS.Timeout | null>(null);
+    const mounted = useRef(false);
+
+    const [isManageMode, setIsManageMode] = useState(false);
+    const [profiles, setProfiles] = useState<UserProfile[]>(EMPTY);
+    const [channelMembers, setChannelMembers] = useState<ChannelMembership[]>(EMPTY_MEMBERS);
+    const [searchResults, setSearchResults] = useState<UserProfile[]>(EMPTY);
+    const [loading, setLoading] = useState(false);
+    const [term, setTerm] = useState('');
+
+    const loadedProfiles = (users: UserProfile[], members: ChannelMembership[]) => {
+        if (mounted.current) {
+            setLoading(false);
+            setProfiles(users);
+            setChannelMembers(members);
+        }
+    };
+
+    const clearSearch = useCallback(() => {
+        setTerm('');
+        setSearchResults(EMPTY);
+    }, []);
+
+    const getProfiles = useCallback(debounce(async () => {
+        const hasTerm = Boolean(term);
+        if (!loading && !hasTerm && mounted.current) {
+            setLoading(true);
+            const options = {sort: 'admin', active: true};
+            const {users, members} = await fetchChannelMemberships(serverUrl, channelId, options, true);
+            if (users.length) {
+                loadedProfiles(users, members);
+            }
+        }
+    }, 100), [channelId, loading, serverUrl, term]);
+
+    const handleSelectProfile = useCallback(async (profile: UserProfile) => {
+        const title = formatMessage({id: 'mobile.routes.user_profile', defaultMessage: 'Profile'});
+        const props = {
+            channelId,
+            closeButtonId: CLOSE_BUTTON_ID,
+            location: USER_PROFILE,
+            manageMode: isManageMode,
+            userId: profile.id,
+            canManageAndRemoveMembers,
+        };
+
+        Keyboard.dismiss();
+        openAsBottomSheet({screen: USER_PROFILE, title, theme, closeButtonId: CLOSE_BUTTON_ID, props});
+    }, [canManageAndRemoveMembers, isManageMode]);
+
+    const searchUsers = useCallback(async (searchTerm: string) => {
+        const lowerCasedTerm = searchTerm.toLowerCase();
+        setLoading(true);
+
+        const options: SearchUserOptions = {team_id: currentTeamId, in_channel_id: channelId, allow_inactive: false};
+        const {data = EMPTY} = await searchProfiles(serverUrl, lowerCasedTerm, options);
+
+        setSearchResults(data);
+        setLoading(false);
+    }, [serverUrl, channelId, currentTeamId]);
+
+    const search = useCallback(() => {
+        searchUsers(term);
+    }, [searchUsers, term]);
+
+    const onSearch = useCallback((text: string) => {
```

Comment:
There is a rare but nasty race condition here. We can consider deferring it to a different PR, since it should not happen often.

Imagine this scenario:
- I write a character
- The search timeout passes, `searchUsers` is called
- The server call get lost in the internet, and also hits the most busy replica because of reasons, taking lots of time
- I add another character
- The search timeout passes, `searchUsers` is called
- This package goes fast as a lightning, and finishes before the previous search call.
- The second call sets the search results, and loading to false.
- The first call finally gets through, and sets the results too.
- The user experience is:
  - For a brief moment I had the list of results for "ab" and suddenly I have the results for only "a"

I think I solved something like this by setting in a reference the time of the "last request to the server". Then, each "search users" call knows when it was called. If I was called when the last request to server was made, that means my information is relevant. If I was called before the last request to the server, I went to the server for nothing, so I can dismiss my results (and not set loading back to false).

Reply:
ticket created for this one https://mattermost.atlassian.net/browse/MM-50037